### PR TITLE
Add analyze_question_response API

### LINKED ISSUE #35: Generation analysis for subsection / question
### Context
During the session, the student (dummy agent) will ask questions guided by the plan (as an objective for the session), when the agent decides that he/she is satisfied with the teacher's (user) explanation or after an amount of time has passed (it will say let's move on with another question), it will call an analyse subsection function in the backend, the teacher will continue explaining for the next question and when the processing is done, there will be a session notification for delayed rewards/penalties (score based and/or life-based), analyze subsection will save all the feedback to session data.

### Task
Takes the transcript for a subsection and analyses the student's response, returning analysis in the following form (along with other metadata about the session and question):

image

### PR DESCRIPTION
**Pull request by thatjosh**
### Code changes
* added code for `analyze_question_response`
* added custom db function `get_transcript_by_question`
* updated constants in constants library
* deleted test.ipynb (experimental file for the API)

### API description
`analyze_question_response` takes a request body and returns an analysis of the user's explanation.

### To do:
* Store the question analysis into the DB

### Data description
Input:
```json
{ "name": "Azure", "user_id":"KenjiPcx", "question_id":2345, "session_id":"test1" }
```

Output:
```json
{
   "clarity_score":"4",
   "clarity_explanation":"Your explanation does an excellent job of breaking down the merge sort process into understandable parts, using a logical flow that builds from dividing the list into merging it back together. You effectively use analogies, like 'it's like weaving two sorted strands together,' which helps in making complex processes more relatable. However, there were instances where the explanation could benefit from slightly more simplification or clearer transitions between ideas to achieve a perfect score.",
   "conciseness_score":"3",
   "conciseness_explanation":"You managed to keep your explanation relatively brief, focusing on key aspects of merge sort. Nevertheless, some parts of the explanation, especially in detailing the merging process with pointers, could be more streamlined. While the detail is appreciated for understanding, conciseness could be improved by condensing some of the descriptions without sacrificing clarity.",
   "comprehensiveness_score":"4",
   "comprehensiveness_explanation":"You covered all the fundamental aspects of merge sort, including its process, efficiency, and comparison with other sorting algorithms. Your inclusion of time complexity and scenarios where merge sort might not be the best option provides a comprehensive view. Minor details or further examples of scenarios where merge sort's memory usage is a significant downside could enhance the comprehensiveness.",
   "correctness_score":"5",
   "correctness_explanation":"The information presented in your explanation is accurate, demonstrating a solid understanding of merge sort, its efficiency, and its comparison to other sorting algorithms. You correctly identified the big O notation for merge sort and appropriately discussed its implications.",
   "exemplification_score":"4",
   "exemplification_explanation":"Your use of examples, such as splitting [3,5,2,1] into individual numbers and the analogy of weaving strands, effectively illustrates the merge sort process. These examples are relevant and enhance understanding. Incorporating more varied examples or hypothetical scenarios could provide even clearer insights into the sorting process and its considerations.",
   "adaptation_score":"4",
   "adaptation_explanation":"You did well to adapt your explanation to an undergraduate level student, using appropriate terminology and complexity. Your analogies and step-by-step breakdown of the process were particularly effective. To further improve, ensuring that every technical term (e.g., 'big O notation', 'in place') is briefly explained could make the explanation accessible to all undergraduates, regardless of their prior knowledge in computer science.",
   "question":"What is XXX",
   "question_id":2345
}
```
image

Seed data used for DB:
```json
{
    "session_id": "test1",
    "user_id": "KenjiPcx",
    "student_persona": "Undergraduate level student",
    "session_transcripts": [
        {
            "question_id": 2345,
            "question": "What is XXX",
            "question_transcript": [
                {
                    "user": "merge sort",
                    "assistant": "(Learner): Oh, merge sort sounds interesting! I've heard it's a way to sort items in a list, but I'm not quite sure how it works. Could you start by explaining what merge sort is and why it's used?"
                },
                {
                    "user": "You are right, it is a way to sort numbers in a list, you start by dividing a list of numbers into new lists until you have lists of just one number. So if you have [3,5,2,1] you would divide this until you have [[3],[5],[2],[1]]. When you do this, you can imagine a tree forming. Essentially as you divide the tree in half, you have 2 lists on the second level, and as you keep dividing, you end up with a list of lists with just 1 item. Then you can fall back onto the previous level by merging the two sorted lists back to the original list, and you do this until you get back the original list.",
                    "assistant": "(Learner): Wow, I think I'm starting to get it! So, we split the list into smaller lists until each one has just one number. This makes me wonder, how do we go from these individual lists back to a full, sorted list? Could you explain a bit more about how the merging process works after splitting the list?"
                },
                {
                    "user": "Suppose you have [[3],[5],[2],[1]], [3] and [5] was obtained by splitting [3,5] into 2 lists, so when we merge them back, we just need to merge these two lists that are sorted. In this case 3<5 so 3 comes before 5 in the new list. You do this for all the lists, and you should get 2 lists of 2 elements, then you will repeat this again, and then you will get the original list. But the second time you merge, you will realise that there are more than 2 elements on each list. But you would realise that they are already sorted up to the point of merging. So what you want to do is to have 2 pointers, one pointing at the start of each list and going through both lists. so that you can merge them into a new list. With these two pointers, you can compare and check which number from the both lists is smaller, then you add that number into the list. But once you've done that, you want to move the pointer away, and keep doing that until you have added all the numbers.",
                    "assistant": "(Learner): Oh, I see! So, by using pointers, we can keep track of which elements we've already considered in each of the smaller, sorted lists, and then we carefully merge these lists together by comparing their elements one by one. It's like weaving two sorted strands together into a single, sorted strand, right? I'm curious, how do we handle the situation where one list runs out of elements before the other? What happens to the remaining elements in the longer list during the merge process?"
                },
                {
                    "user": "Yeap, so when that happens you just add the rest of the list that is not being added to the list because at that pointer these remaining numbers are just smaller than all the added numbers, and remember they are sorted so it works",
                    "assistant": "(Learner): Oh, that makes sense! Because the lists we are merging are already sorted, if one list runs out of elements first, we can simply append the remaining elements from the other list directly to the end of the merged list, knowing they're in the correct order. It's like adding the last pieces to complete a puzzle. This process of dividing, sorting individually, and then merging back together ensures the entire list is sorted in the end. I think I understand the basic idea of merge sort now! It's a very systematic way to sort a list. Is there a specific reason why merge sort is preferred over other sorting methods in certain situations?"
                },
                {
                    "user": "Yes, merge sort is more efficient than the other sorting algorithms like insertion sort or selection sort. Formally this is known as time complexity, and merge sort has a big O of nlogn, as opposed to n^2 for the other sorting algorithms.",
                    "assistant": "(Learner): Ah, time complexity! That's about how long an algorithm takes to complete its task based on the size of its input, right? So, if merge sort has a time complexity of \\(O(n \\log n)\\), it means that as the list of numbers gets bigger, merge sort can sort the list more quickly than those other sorts with a time complexity of \\(O(n^2)\\), especially when the lists are very large. It's fascinating how the way we sort can make such a big difference in efficiency. This makes me wonder, are there scenarios where merge sort might not be the best choice, even with its impressive time complexity?"
                },
                {
                    "user": "Merge sort does not sort the numbers in place, meaning you need to create new lists which requires memory. If you don't want to use memory then you might want to consider heapsort, which has the same time complexity but it does not use extra memory",
                    "assistant": "(Learner): Oh, I see! So, while merge sort is efficient in terms of time, it requires extra memory because it has to create new lists during the sorting process. This trade-off between time efficiency and memory usage is really important to consider, especially in situations where memory is limited. Heapsort, on the other hand, sorts the list in place, saving memory but still maintaining that efficient \\(O(n \\log n)\\) time complexity. It's like choosing between two tools, where each has its own advantages depending on the job at hand. This really highlights the importance of understanding different algorithms and their characteristics. Thanks for explaining merge sort and its considerations so clearly! Is there another topic or concept you'd like to explore or explain?"
                }
            ]
        },
        {
            "question_id": 23456,
            "question": "What is XXX",
            "question_transcript": [
                {
                    "user": "T1",
                    "assistant": "T2"
                }
            ]
        }
    ],
}
```

### CODE REVIEW

File: api/constants.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 # Library for storing global constants
 
+from langchain.output_parsers import ResponseSchema
+
 # Marking rubric for a learner's explanation
-marking_rubric = """
+MARKING_RUBRIC = """
```
Comment (KenjiPcx) [inline api/constants.py:6]: Why do we only have a rubric for clarity? We can include one about correctness and conciseness also, prompt can be a little shorter

File: api/function_app.py [hunk 2]
```diff
@@ -36,6 +37,7 @@
 openai_key = os.getenv("OPENAI_API_KEY")
 openai_client = openai.OpenAI(api_key=openai_key)
 langchain_llm = ChatOpenAI(api_key=openai_key, model="gpt-4-turbo-preview")
+chat_model = ChatOpenAI(model="gpt-4-turbo-preview", openai_api_key=openai_key)
```
Comment (KenjiPcx) [inline api/function_app.py]: duplicated variable

**Review by KenjiPcx — CHANGES_REQUESTED** [review]

File: api/constants.py (hunk 1 above)
Comment (KenjiPcx) [reply api/constants.py:6]: uhh this one can ignore first la

File: api/function_app.py (hunk 2 above)
Comment (thatjosh) [reply api/function_app.py]: mb, will remove

File: api/constants.py (hunk 1 above)
Comment (thatjosh) [reply api/constants.py:6]: The rest of the criteria is hidden, check the entire file for the rest of the criteria

**Pull request merged**